token/inc/token.h

```diff
@@ -0,0 +1,240 @@
+     *
+     * # Accounts expected by this instruction:
+     *
+     *   0. `[signer]` Owner of the source account.
+     *   1. `[]` Source account.
+     *   2. `[writable]` Delegate account.
+     */
+    Approve,
+    /**
+     * Sets a new owner of an account.
+     *
+     * # Accounts expected by this instruction:
+     *
+     *   0. `[signer]` Current owner of the account.
+     *   1. `[writable]` account to change the owner of.
+     *   2. `[]` New owner of the account.
+     */
+    SetOwner,
+    /**
+     * Mints new tokens to an account.
+     *
+     * # Accounts expected by this instruction:
+     *
+     *   0. `[signer]` Owner of the token.
+     *   1. `[writable]` Token to mint.
+     *   2. `[writable]` Account to mint tokens to.
+     */
+    MintTo,
+    /**
+     * Burns tokens by removing them from an account and the total supply.
+     *
+     * # Accounts expected by this instruction:
+     *
+     *   0. `[signer]` Owner of the account to burn from.
+     *   1. `[writable]` Account to burn from.
+     *   2. `[writable]` Token being burned.
+     */
+    Burn,
+} Instruction_Tag;
+
+typedef struct NewToken_Body {
+    TokenSupply _0;
+} NewToken_Body;
+
+typedef struct Transfer_Body {
+    uint64_t _0;
+} Transfer_Body;
+
+typedef struct Approve_Body {
+    uint64_t _0;
+} Approve_Body;
+
+typedef struct MintTo_Body {
+    uint64_t _0;
+} MintTo_Body;
+
+typedef struct Burn_Body {
+    uint64_t _0;
+} Burn_Body;
+
+typedef struct Instruction {
+    Instruction_Tag tag;
+    union {
+        NewToken_Body new_token;
+        Transfer_Body transfer;
+        Approve_Body approve;
+        MintTo_Body mint_to;
+        Burn_Body burn;
+    };
+} Instruction;
+
+typedef uint8_t Pubkey[32];
+
+/**
+ * Represents a token type identified and identified by its public key.  Accounts
+ * are associated with a specific token type and only accounts with
+ * matching types my inter-opt.
+ */
+typedef struct TokenInfo {
+    /**
+     * The total supply of tokens.
+     */
+    TokenSupply info;
+    /**
+     * True if this token has an owner
+     */
+    bool owner_present;
+    /**
+     * Optional token owner that is ignored if `owner_present` is false, used to mint new tokens.
+     * The owner may only be provided during token creation.  If no owner is present then the
+     * token has a fixed supply and no further tokens may be minted.
+     */
+    Pubkey owner;
+} TokenInfo;
+
+/**
+ * Delegation details.
+ */
+typedef struct AccountDelegate {
+    /**
+     * The source account for the tokens.
+     */
+    Pubkey source;
+    /**
+     * The original maximum amount that this delegate account was authorized to spend.
+     */
+    uint64_t original_amount;
+} AccountDelegate;
+
+/**
+ * Account that holds or may delegate tokens.
+ */
+typedef struct TokenAccountInfo {
+    /**
+     * The type of token this account holds.
+     */
+    Pubkey token;
```

Comment:
Do we have control over whether members are pointers or not? Since the Ledger S is memory constrained, we point into the original TX message buffer instead of copying for most types that are larger than a `u64`

Reply:
That structure is not part of the instruction, its the token program's account data.  Also, any reason to not just cast a pointer to the account data to a pointer to this struct?